docs_src/design/adr/014-Secret-Provider-For-All.md

```diff
@@ -0,0 +1,385 @@
+	LogLevel        string
+	...
+	InsecureSecrets InsecureSecrets
+}
+
+type InsecureSecrets map[string]InsecureSecretsInfo
+
+type InsecureSecretsInfo struct {
+	Path    string
+	Secrets map[string]string
+}
+```
+
+```toml
+[Writable.InsecureSecrets]
+    [Writable.InsecureSecrets.DB]
+    	path = "redisdb"
+    	[Writable.InsecureSecrets.DB.Secrets]
+            username = ""
+            password = ""
+    [Writable.InsecureSecrets.mqtt]
+    	path = "mqtt"
+    	[Writable.InsecureSecrets.mqtt.Secrets]
+            username = ""
+            password = ""
+            cacert = ""
+            clientcert = ""
+            clientkey = ""
+```
+
+## Decision
+
+The new `SecretProvider` abstraction defined by this ADR is a combination of the two implementations described above in the [Existing Implementations](#existing-implementations) section.
+
+### Only Exclusive Secret Stores
+
+To simplify the `SecretProvider` abstraction, we need to reduce to using only exclusive `SecretStores`. This allows all the APIs to deal with a single `SecretClient`, rather than the split up way we currently have in Application Services. This requires that the current Application Service shared secrets (database credentials) must be copied into each Application Service's exclusive `SecretStore` when it is created.
+
+The challenge is how do we seed known secrets for unknown services when they become known.  As described above in the [Known and Unknown Services](#known-and-unknown-services) section above,  services currently identify themselves for exclusive `SecretStore` creation via the `ADD_SECRETSTORE_TOKENS` environment variable on security-secretstore-setup. This environment variable simply takes a comma separated list of service names.
+
+```yaml
+ADD_SECRETSTORE_TOKENS: "<service-name1>,<service-name2>"
+```
+
+If we expanded this to add an optional list of known secret identifiers for each service, i.e. `Databases.application-service`, the exclusive store could also be seeded with a copy of known shared secrets. In this case the database credentials for the Application Services' shared database. The environment variable name will change to `ADD_SECRETSTORE` now that it is more than just tokens.
+
+```yaml
+ADD_SECRETSTORE: "app-service-xyz[Databases.application-service]"
+```
+
+Full specification for the environment variable's value is a comma separated list of service entries defined as:
+
+```
+<service-name>[optional list of known secret IDs sperated by ;]>
+```
+
+Example with one service specifying IDs for know secrets and one without known secrets 
+
+```yaml
+ADD_SECRETSTORE: "appservice-xyz[Databases.application-service; <other TBD id>], appservice-http-export"
+```
+
+Once security-secretstore-setup generates the known application-service database credentials, it will save them in memory while processing additional adds of `SecretStores` so they can be used if other Application Services needs them seeded into their `SecretStores` . 
```

Comment:
## What actually happens today

1. security-secretstore-setup spawns a separate executable, security-file-token-provider.
1. the file-token-provider enumerates a configuration file and creates a policy and token for each service using its service key.  This token give access its the services' private area, e.g. secrets/edgex/edgex-core-data, as well as any custom paths carried over for legacy purposes, e.g. secret/edgex/coredata/mongodb
1. later on, we create database credentials by enumerating the configuration.toml and create a database secret for each database where a Service= key is specified.  For example: secret/edgex/coredata/mongodb  when Service=coredata.  There is also a copy of the secrets under secret/edgex/mongo or under secret/edgex/bootstrap-redis.  For mongodb, it uses the section name, such as application-service instead of the service tag, appservice. secret/edgex/mongo/application-service contains the same secret as secret/edgex/appservice/mongodb.  The reason for the duplication is that the database bootstrapper can enumerate the entire secret/edgex/mongo subtree when provisioning accounts.

## What will happen in the bootstrapping ADR

Database credentials will be bootstrapped as part of the install.  security-file-token-provider will run every boot to regenerate tokens.

## What this means for the above

security-file-token-provider doesn't really know anything about databases.  It just knows about tokens and policies.  Similarly, for legacy reasons the old code isn't even putting the database secrets in the right place to begin with -- where is inconsistent use of the service key in the framework.  (continue later)

Reply:
@bnevis-i,  I like where you are going. I think we can drop `duplicate` since any secret specified here is to be duplicated. Also since `/secret/edgex/` is always present, we could drop it also. This would make it nice and concise as:
 ```
app-service-xyz[appservice/redisdb]
```
If we also have credentials for message bus if would look something like:
```
app-service-xyz[appservice/redisdb; common/messagebus]
```
This would create the following for `edgex-app-service-xyz`
```
/secret/edgex/edgex-app-service-xyz/redisdb
/secret/edgex/edgex-app-service-xyz/messagebus
```
and assumes the following new policy add to `token-config-json` :
```
  "edgex-service-common": {
    "edgex_use_defaults": true,
    "custom_policy": {
      "path": {
        "secret/edgex/common/messagebus": {
          "capabilities": [
            "list",
            "read"
          ]
        },
      }
    }
  },
```

This all depends on the database or other credential bootstrapping have completed and stored the credentials before the environment variable is processed. Don't think it happening in that order now, but we could refactor to ensure it is. Thoughts?

Reply:
This makes sense.

However, `token-config.json` is just for creating tokens.  `token-config.json` doesn't need to be altered as per above unless one is creating a `service-common` service that needs its own token.  All of the known secret seeding today is done with a root token that has access to the entire KV store.

Reply:
@bnevis-i , My thoughts with `service-common` is it is similar to `appservice`. There isn't a specific service called AppService, but we create the token and secret store for it, put the app service shared secrets in it and then use it to copy the shared secrets to actual App Services secret stores when they declare themselves via the environment variable.

`service-common` would be there for shared secrets that all edgex services need, such as the message bus credentials. Does this make sense?

Reply:
I'm not opposed to the idea of `service-common`.  I am just suggesting that we might not need a separate token for it.  security-secretstore-setup is trusted to read/write secrets anywhere in the URL space, whether it is covered by a token or not.